### PR TITLE
Implement ring buffer for head tracker that drops oldest head

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

File: core/services/head_tracker.go [hunk 1]
```diff
@@ -48,17 +48,64 @@ var (
...
 
 const (
-	// The size of the buffer for the headers channel. Note that callback
-	// execution is synchronous and could take a non-trivial amount of time.
-	headsBufferSize = 5
+	// The size of the buffer for the headers channel. This number controls
+	// how many heads we will buffer before dropping. Setting this to a number
+	// greater than 0 allows some short periods of overload at the expense of
+	// lagging processing.
+	HeadsBufferSize = 3
```
Comment (kaleofduty) [inline core/services/head_tracker.go:55]: Shouldn't this be configurable?

**Comment by samsondav** [comment]
I think I might rewrite this as a small data structure with two goroutines and a mutex, since it could be made more correct and easier to reason about.

File: core/services/head_tracker.go [hunk 2]
```diff
@@ -42,23 +42,83 @@ var (
 		Help:    "How long it took to execute all callbacks (ms) histogram",
 		Buckets: []float64{50, 100, 250, 500, 1000, 2000, 5000, 10000, 15000, 30000, 100000},
 	})
+	promNumHeadsDropped = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "head_tracker_num_heads_dropped",
+		Help: "The total number of heads dropped",
+	})
```
Comment (samsondav) [inline core/services/head_tracker.go:48]: NOTE: This needs an alert added here: https://github.com/smartcontractkit/infra/blob/master/tools/cortex/mgnt/staging/config/ec2/chainlink.rules.yml

**Comment by samsondav** [comment]
@se3000 https://www.pivotaltracker.com/story/show/174598974

**Review by topliceanu — APPROVED** [review]

**Pull request merged**